evm/db/chain.py

```diff
@@ -131,6 +169,8 @@ def add_block_number_to_hash_lookup(self, header):
             rlp.encode(header.hash, sedes=rlp.sedes.binary),
         )
 
+    # TODO: This method sould take a chain of headers as that's the most common use case
+    # and it'd be much faster than inserting each header individually.
```

Comment:
What sort of optimizations do you think are likely to make this faster if this took a list of headers?

Reply:
For every header we insert we check if it should be the head of the canonical chain, and if it does we have to go through its ancestors to find the chain leading up to the first one it has in common with the canonical chain, then update the block-number-to-hash-lookup entries in the DB for them. If we get a chain of headers here we would be able to do the should-be-canonical check *and* lookup-chain-up-to-common-ancestor just once instead of for every header.

Reply:
Ah, yes, that makes sense.